src/ripple/overlay/impl/Cluster.cpp

```diff
@@ -0,0 +1,136 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/basics/Log.h>
+#include <ripple/core/Config.h>
+#include <ripple/core/TimeKeeper.h>
+#include <ripple/overlay/Cluster.h>
+#include <ripple/overlay/ClusterNode.h>
+#include <ripple/protocol/JsonFields.h>
+#include <ripple/protocol/RippleAddress.h>
+#include <ripple/protocol/tokens.h>
+#include <boost/regex.hpp>
+#include <memory.h>
+
+namespace ripple {
+
+Cluster::Cluster (beast::Journal j)
+    : j_ (j)
+{
+}
+
+boost::optional<std::string>
+Cluster::member (RippleAddress const& identity)
+{
+    std::lock_guard<std::mutex> _(lock_);
+
+    auto iter = nodes_.find (identity);
+    if (iter == nodes_.end ())
+        return boost::none;
+    return iter->name ();
+}
+
+bool
+Cluster::update (
+    RippleAddress const& identity,
+    std::string name,
+    std::uint32_t loadFee,
+    std::uint32_t reportTime)
+{
+    std::lock_guard<std::mutex> _(lock_);
+
+    // NIKB FIXME: We can't use auto here yet due to an
+    //             issue with libstdc++.
+    std::set<ClusterNode, Comparator>::iterator iter =
+        nodes_.find (identity);
+
+    if (iter != nodes_.end ())
+    {
+        if (reportTime <= iter->getReportTime())
+            return false;
+
+        if (name.empty())
+            name = iter->name();
+
+        iter = nodes_.erase (iter);
+    }
+
+    nodes_.emplace_hint (iter, identity, name, loadFee, reportTime);
+    return true;
+}
+
+void
+Cluster::for_each (
+    std::function<void(ClusterNode const&)> func)
+{
+    std::lock_guard<std::mutex> _(lock_);
```

Comment:
We've this this pattern cause problems in OverlayImpl. Since both `for_each` and `update` are public functions, we can't guarantee that `func` won't call `update` and invalidate our iterator. Maybe a bool that says when we're locked, set to true in for_each and at least assert(!locked) in `update`? Doesn't need to be atomic since the "bad" case is all in one thread. (I don't know what happens when a thread locks a mutex it already owns. The standard just says it "must not do this").